src/jupyterhub_config/custom_kube_spawner.py

```diff
@@ -0,0 +1,345 @@
+        self.environment['HOME'] = str(user_dir)
+        self.environment['PATH'] = f"{user_env_dir}/bin:{os.environ['PATH']}"
+        self.environment['VIRTUAL_ENV'] = str(user_env_dir)
+        if 'PYTHONPATH' in os.environ:
+            self.environment['PYTHONPATH'] = f"{user_env_dir}/lib/python3.11/site-packages:{os.environ['PYTHONPATH']}"
+        else:
+            self.environment['PYTHONPATH'] = f"{user_env_dir}/lib/python3.11/site-packages"
+        self.environment['PYTHONSTARTUP'] = os.path.join(os.environ['JUPYTERHUB_CONFIG_DIR'], 'startup.py')
+        self.environment['JUPYTERHUB_USER'] = username
+        self.environment['SHELL'] = '/bin/bash'
+
+        if self._is_rw_minio_user():
+            self.log.info(f"MinIO read/write user detected: {self.user.name}. Setting up minio_rw credentials.")
+            self.environment['MINIO_ACCESS_KEY'] = self.environment['MINIO_RW_ACCESS_KEY']
+            self.environment['MINIO_SECRET_KEY'] = self.environment['MINIO_RW_SECRET_KEY']
+            self.environment['USAGE_MODE'] = 'dev'
+        else:
+            self.log.info(f"Non-admin user detected: {self.user.name}. Removing admin credentials.")
+            self.environment.pop('MINIO_RW_ACCESS_KEY', None)
+            self.environment.pop('MINIO_RW_SECRET_KEY', None)
+
+        self.environment.pop('JUPYTERHUB_ADMIN_PASSWORD', None)
+        self.log.info(f"Environment variables for user '{self.user.name}' at pod startup: {self.environment}")
+
+    def _configure_notebook_dir(self, username: str, user_dir: Path):
+        """
+        Configure the notebook directory for the user. If the user is an admin,
+        the directory is set to a shared workspace. Otherwise, it is set to the
+        user's home directory.
+        """
+        if self.user.admin:
+            self.log.info(f'Admin user detected: {username}. Setting up admin workspace.')
+            # root directory
+            self.notebook_dir = str('/')
+        else:
+            self.log.info(f'Non-admin user detected: {username}. Setting up user-specific workspace.')
+            # TODO: It appears that notebook_dir must be the parent of the favorites directory - investigate if it's possible to set notebook_dir to user_dir
+            # self.notebook_dir = str(user_dir)
+            self.notebook_dir = str('/')
+
+    def _is_rw_minio_user(self):
+        """
+        Check if the user is a read/write MinIO user.
+
+        Admin users and users in the minio_rw group are considered read/write MinIO users.
+        """
+        group_names = [group.name for group in self.user.groups]
+        return self.user.admin or self.RW_MINIO_GROUP in group_names
+
+    def _ensure_user_volume(self, user_dir: Path):
+        """
+        Ensure the user's volume is correctly mounted in the container.
+        """
+
+        user_home_dir = os.environ['JUPYTERHUB_USER_HOME']
+        mount_base_dir = os.environ.get('JUPYTERHUB_MOUNT_BASE_DIR')
```

Comment:
Under what situations will this not be available?

Reply:
it should never be none. Probably a copy error.

Reply:
Can we revert to square brackets then?

Reply:
:+1: